### ISSUE TITLE
how to reduce latency of tcpip forward

### ISSUE THREAD

**Issue by yangsengui** [mscdex/ssh2#1285]
I use ssh2 to proxy tcpip to the internal host, so that jetbrains gateway can connect to the internal host.
local -> ssh2 server(like a gateway, proxy the tcpip connection) -> internal host

the latency
local -> ssh2 server      about 35ms(from ping)
ssh2 server -> internal host     it's negligible

i think 50ms to the internal host can be accepted, but the latency of connection to internal host is unstable, the latency range from 46ms even to 130 ms.

I guess it's beacuse of the property highWaterMark. I try ot set it  at1000 and 1 ,it has no different.

![image](https://user-images.githubusercontent.com/59313293/229327337-04b7cac7-6abb-4ffe-8fe2-2fa134f5b34d.png)

here is my ssh2 server code
```
const { timingSafeEqual } = require('crypto');
const { readFileSync } = require('fs');
const { inspect } = require('util');

const { utils: { parseKey }, Server, Client } = require('ssh2');

const allowedUser = Buffer.from('foo');
const allowedPassword = Buffer.from('bar');
const allowedPubKey = parseKey(readFileSync('C:\\Users\\yangs\\.ssh\\id_rsa'));
const conn = new Client();
conn.connect({
    host: '127.0.0.1',
    port: 22,
    username: "root",
    password: "123",
        debug: (e) => {
    // console.log(e)
}
},);
conn.on("error", (err) => {
    console.log(err)
})
function checkValue(input, allowed) {
    const autoReject = (input.length !== allowed.length);
    if (autoReject) {
        // Prevent leaking length information by always making a comparison with the
        // same input when lengths don't match what we expect ...
        allowed = input;
    }
    const isMatch = timingSafeEqual(input, allowed);
    return (!autoReject && isMatch);
}

new Server({
    hostKeys: [readFileSync('C:\\Users\\yangs\\.ssh\\id_rsa')],
    highWaterMark: 1000,
    debug: (e) => {
    // console.log(e)
}
}, (client) => {
    client.on('authentication', (ctx) => {
        let allowed = true;
        if (!checkValue(Buffer.from(ctx.username), allowedUser))
            allowed = false;
        switch (ctx.method) {
            case 'password':
                if (!checkValue(Buffer.from(ctx.password), allowedPassword))
                    return ctx.reject();
                break;
            case 'publickey':
                if (ctx.key.algo !== allowedPubKey.type
                    || !checkValue(ctx.key.data, allowedPubKey.getPublicSSH())
                    || (ctx.signature && allowedPubKey.verify(ctx.blob, ctx.signature) !== true)) {
                    return ctx.reject();
                }
                break;
            default:
                return ctx.reject(["password"]);
        }

        if (allowed)
            ctx.accept();
        else
            ctx.reject();
    }).on('ready', () => {
        client.on("request", (accept, reject, name, info) => {
            accept()
        })
        client.on("tcpip", (accept, reject, info) => {
            const stream = accept()
            conn.forwardOut(info.srcIP, info.srcPort, info.destIP, info.destPort, (err, clientStream) => {
                clientStream.pipe(stream).pipe(clientStream)
            })
        })

        client.on('session', (accept, reject) => {
            const session = accept();

            session.once('shell', function(accept, reject, info){
                conn.shell(info, {}, function(err, stream){
                    if(err) return reject()
                    const a = accept()
                    a.pipe(stream).pipe(a)
                })
            })
            session.once('exec', (accept, reject, clientInfo) => {
                const stream = accept();
                conn.exec(clientInfo.command, (err, clientStream) => {
                    clientStream.pipe(stream, { end: false }).pipe(clientStream)
                    clientStream.stderr.pipe(stream.stderr);
                    clientStream.on("exit", (code, signal) => {
                        stream.exit(code)
                        stream.end()
                    })
                })
            });
            session.once('pty', function(accept, reject, info){
                accept && accept()
            })
            session.once('sftp', function(accept, reject){
                if (accept) {
                    accept()
                }
            })
        });
    }).on('close', () => {
        console.log("disconnected")
    });
}).listen(8023, '0.0.0.0', function() {
    console.log('Listening on port ' + this.address().port);
}).on("error", (err) => {
    console.log(err)
})```

**Comment by mscdex** [comment]
I don't have any suggestions. You might measure the latency between between the client and the `ssh2` server to find out if the latency issues are between the client and `ssh2` server or the `ssh2` server and the upstream server you're proxying to.

Changing the `highWaterMark` for any of the involved streams is not going to affect latency, only the amount of memory buffered before backpressure kicks in.

You could also compare it with something like an OpenSSH server.

Either way, I'm not sure there's much that can be done in the way of improving latency.

**Comment by yangsengui** [comment]
https://nodejs.org/api/net.html#netcreateserveroptions-connectionlistener
I found a option of net.Server noDelay, could it be useful?

**Comment by mscdex** [comment]
Feel free to fork the project, make the required changes for enabling `noDelay` and test it out to see if it makes any meaningful difference for you (it may not). If it does, I can look into adding support for it.

**Comment by yangsengui** [comment]
I did some test, noDelay option is helpful in improve latency, not just in tcip forward , shell can also benefit. I think it is an important option for those who care about the latency of connection.

I did a littile change in server.js, just one line. I think a screenshot can help you to add support for it.
![image](https://user-images.githubusercontent.com/59313293/229349370-10d15597-cdd5-4c01-9092-fad3c4f5e10b.png)
